### PR TITLE
discovery: Get injected environment variables

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?
   
Get injected environment variables from the auto injector (Datadog/auto_inject) via a memfd which the auto injector creates and makes available with a specific name via /proc/$PID/fds.

This PR is based on https://github.com/DataDog/datadog-agent/pull/28290.

Note that the PR to add the feature to Datadog/auto_inject is not merged yet: https://github.com/DataDog/auto_inject/pull/342

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1156

### CODE REVIEW

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+	return envMap
+}
+
+// getEnvs gets the environment variables for the process, both the initial
+// ones, and if present, the ones injected via the auto injector.
+func getEnvs(proc *process.Process) (map[string]string, error) {
+	envs, err := proc.Environ()
+	if err != nil {
+		return nil, nil
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:94]:
```suggestion
		return nil, err
```

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+			continue
+		}
+
+		if name != injectorMemFdPath {
+			continue
+		}
+
+		envs, _ := readEnvsFile(path)
+		return envs
+	}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:68]: Why can't you access the file directly? Why do we need to iterate over the directory?

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:5]: shouldn't this file be accessible only for linux?

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 2 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:68]: The file names are fd numbers so we don't know which one we want unless we iterate and read the symbolic link targets.

`$ ls -l /proc/1464207/fd
total 0
lrwx------ 1 vincent vincent 64 Aug 13 13:45 0 -> /dev/pts/6
lrwx------ 1 vincent vincent 64 Aug 13 13:45 1 -> /dev/pts/6
lrwx------ 1 vincent vincent 64 Aug 13 13:45 2 -> /dev/pts/6
lrwx------ 1 vincent vincent 64 Aug 13 13:45 3 -> '/memfd:foo (deleted)'
`

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 3 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:5]: I've kept it the same as `impl_linux.go`, which doesn't have that.

Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:5]: if the file name ends with `_linux.go` (or `_windows.go`) it implies `linux` build only

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 2 above)
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:68]: so please document that in the code

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 3 above)
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:5]: This reflects in the job failure https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/603129477

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 2 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:84]: Done.

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 3 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:5]: TIL. Done.

**Review by guyarb — APPROVED** [review]
👏🏻

**Pull request merged**